### PR TITLE
CI: linting check to ensure `lib.NoDefault` is only used for typing

### PR DESCRIPTION
**Pull request by Charlie-XIAO**
- [x] See https://github.com/pandas-dev/pandas/pull/53877#issuecomment-1609917156
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit)
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions

In this PR, I added a pre-commit hook to check that `pandas._libs.lib.NoDefault` is used only for typing. This is invoked by #53877. Before that PR, this new hook correctly catches all misused `NoDefault`:

<details>
<p></p>

```
(pandas-dev) D:\>pre-commit run unwanted-patterns-nodefault-not-used-for-typing --all-files
Check for pandas._libs.lib.NoDefault not used for typing.......................Failed   
- hook id: unwanted-patterns-nodefault-not-used-for-typing
- exit code: 1

pandas/tests/reshape/test_pivot_multilevel.py:38:NoDefault is not used for typing       
pandas/tests/reshape/test_pivot_multilevel.py:75:NoDefault is not used for typing       
pandas/core/frame.py:8831:NoDefault is not used for typing
pandas/core/frame.py:8831:NoDefault is not used for typing
pandas/core/reshape/pivot.py:514:NoDefault is not used for typing
pandas/core/reshape/pivot.py:515:NoDefault is not used for typing
pandas/core/reshape/pivot.py:529:NoDefault is not used for typing
pandas/core/reshape/pivot.py:525:NoDefault is not used for typing
pandas/core/reshape/pivot.py:530:NoDefault is not used for typing
pandas/core/reshape/pivot.py:535:NoDefault is not used for typing
pandas/core/reshape/pivot.py:542:NoDefault is not used for typing
pandas/core/reshape/pivot.py:572:NoDefault is not used for typing
```

</details>

After that PR, the check passes normally:

<details>
<p></p>

```
(pandas-dev) D:\>pre-commit run unwanted-patterns-nodefault-not-used-for-typing --all-files
Check for pandas._libs.lib.NoDefault not used for typing.......................Passed
```

</details>

In fact I don't work with ASTs very often so not sure if this check covers all possibilties. Please let me know if any modification is needed. (The runtime is a bit long but seems acceptable, approximately 10s to check all files.)

### CODE REVIEW

**Comment by Charlie-XIAO** [comment]
Thanks for the code pointer! I'll look into it.

**Comment by Charlie-XIAO** [comment]
Improved from approx. 10s for all files to 3~4s @MarcoGorelli

**Comment by MarcoGorelli** [comment]
Well done!

Would appreciate it if we could add a tiny little test case to scripts/tests/test_validate_unwanted_patterns.py too

Other than that, looks good!

**Comment by Charlie-XIAO** [comment]
Didn't notice that there are tests for linting scripts, thanks for the reminder. Added some simple tests now.

**Review by MarcoGorelli — APPROVED** [review]
nice, great work, looks good to me on green

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -339,6 +339,11 @@ repos:
         language: python
         entry: python scripts/validate_unwanted_patterns.py --validation-type="strings_with_wrong_placed_whitespace"
         types_or: [python, cython]
+    -   id: unwanted-patterns-nodefault-not-used-for-typing
+        name: Check for `pandas._libs.lib.NoDefault` not used for typing
```
Comment (mroeschke) [inline .pre-commit-config.yaml:343]: Just to clarify `pandas._libs.lib.NoDefault` should **only** be used for typing

Comment (MarcoGorelli) [reply .pre-commit-config.yaml:343]: "Check for `pandas._libs.lib.NoDefault` is only used for typing"

Comment (mroeschke) [reply .pre-commit-config.yaml:343]: Yeah the script looks correct, just a suggestion that this phrasing was a little unclear

Comment (MarcoGorelli) [reply .pre-commit-config.yaml]: If we're going to rename this, can we rename it everywhere please? there's a few places where it still says "not used in typing" - let's use consistent wording, and for the error message, let's say "used not only in typing"

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Nice thanks @Charlie-XIAO

**Pull request merged**